### PR TITLE
#135 Extract actions Core & API.

### PR DESCRIPTION
**Pull request by tomaszmichalak**
Extract actions to separate modules (both Core and API).

## Upgrade notes (if appropriate)

| Old        | New           | 
| ------------- |-------------| 
| `io.knotx.fragments.handler.api.actionlog.*`     | `io.knotx.fragments.action.api.log.*`  |
| `io.knotx.fragments.handler.api.* `     | `io.knotx.fragments.action.api.*`  |
| `io.knotx.fragments.task.factory.ActionFactoryOptions`     | `io.knotx.fragments.action.ActionFactoryOptions`  |
| `io.knotx.fragments.task.factory.node.action.ActionProvider `     | `io.knotx.fragments.action.ActionProvider`  |
| `io.knotx.fragments.task.TaskFactory`     | `io.knotx.fragments.handler.api.task.TaskFactory`  |

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] My code follows the [code style](/CONTRIBUTING.md#coding-conventions) of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

---
I hereby agree to the terms of the Knot.x Contributor License Agreement.

### CODE REVIEW

File: actions/core/src/main/java/io/knotx/fragments/action/ActionProvider.java [hunk 1]
```diff
@@ -49,6 +52,19 @@ public ActionProvider(Supplier<Iterator<ActionFactory>> supplier,
     this.cache = new HashMap<>();
   }
 
+  /**
+   * Use SPI to load {@link ActionFactory}s available on the classpath.
+   */
+  public ActionProvider(Map<String, ActionFactoryOptions> actionNameToOptions, Vertx vertx) {
```
Comment (marcinus) [inline actions/core/src/main/java/io/knotx/fragments/action/ActionProvider.java]: Could call the constructor above
```
super(() -> ServiceLoader.load(ActionFactory.class).iterator(), actionNameToOptions, vertx)
```

File: actions/api/docs/asciidoc/dataobjects.adoc [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+
+[cols=">25%,25%,50%"]
+[frame="topbot"]
+|===
+^|Name | Type ^| Description
+|[[config]]`@config`|`Json object`|+++
+Sets <code>Action</code> configuration that is passed to Action.
++++
+|[[doAction]]`@doAction`|`String`|+++
+Sets the name of the base <code>Action</code> that will be triggered while creating current <code>Action</code>. In not set (<code>null</code>), given action will have no base actions.
```
Comment (marcinus) [inline actions/api/docs/asciidoc/dataobjects.adoc]: This is not really accurate - the `doAction` will not be _triggered_ when _creating current Action_ - but when the current Action is _called_. Would be nice to change even though not introduced in this PR :)

File: actions/README.md [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+Action may be decorated with [behaviours](https://github.com/Knotx/knotx-fragments/tree/master/action/core#behaviours). 
+Behaviour is a special action that wraps the original action and adds some functionality e.g 
+employ the circuit breaker pattern. 
+
+Please note that actions can be **stateless** and **stateful**.
+
+### Knot
+**Knot** is a scalable **Action** that is available on the [Vert.x Event Bus](https://vertx.io/docs/vertx-core/java/#event_bus). 
+See the [Template Engine](https://github.com/Knotx/knotx-template-engine/blob/master/core/src/main/java/io/knotx/te/core/TemplateEngineKnot.java) 
+module as reference.
```
Comment (marcinus) [inline actions/README.md]: > as reference
as an example

File: actions/README.md [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+Action factory options contains:
+- [action factory](#action-factory) name
+- action JSON configuration
+- `doAction` for behaviours
+
+and allows:
+- connecting a action factory with configuration
+- linking actions with behaviours.
+
+For more details [here](https://github.com/Knotx/knotx-fragments/blob/master/action/core/docs/asciidoc/dataobjects.adoc#actionfactoryoptions).
```
Comment (marcinus) [inline actions/README.md]: I think it would sound nice like `For more details look [here]`, or maybe just `More details [here]`?

File: actions/README.md [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    factory = http
+    config.endpointOptions {
+      path = "/service/user/{params.id}"
+      domain = localhost
+      port = 1024
+      allowedRequestHeaders = ["Content-Type"]
+    }
+  }  
+  ```
+- use the [action provider](#action-provider) to and invoke the action e.g.
```
Comment (marcinus) [inline actions/README.md]: to retrieve/get?

File: actions/README.md [hunk 6]
```diff
@@ -0,0 +1,102 @@
+# Fragments Actions
+It is a tool-kit for integrating with different data sources, especially with HTTP-based APIs. It
```
Comment (Skejven) [inline actions/README.md]:
```suggestion
It is a tool-kit for integrating with different data sources (e.g. HTTP-based APIs). It 
```

File: actions/README.md [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+- connecting a action factory with configuration
+- linking actions with behaviours.
+
+For more details [here](https://github.com/Knotx/knotx-fragments/blob/master/action/core/docs/asciidoc/dataobjects.adoc#actionfactoryoptions).
+
+### Action factory
+Action [factory](https://github.com/Knotx/knotx-fragments/blob/master/handler/api/src/main/java/io/knotx/fragments/handler/api/ActionFactory.java)
+creates an action instance providing an [action configuration](#action-configuration). It is identified by name (a simple text value).
+
+A list of available action factories [here](https://github.com/Knotx/knotx-fragments/tree/master/action/core).
```
Comment (Skejven) [inline actions/README.md]:
```suggestion
See the full list of available action factories [here](https://github.com/Knotx/knotx-fragments/tree/master/action/core).
```

File: actions/README.md [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+
+### Knot
+**Knot** is a scalable **Action** that is available on the [Vert.x Event Bus](https://vertx.io/docs/vertx-core/java/#event_bus). 
+See the [Template Engine](https://github.com/Knotx/knotx-template-engine/blob/master/core/src/main/java/io/knotx/te/core/TemplateEngineKnot.java) 
+module as reference.
+
+### Action provider
+Action provider initializes actions, combines actions with behaviours and caches stateful ones. It
+uses aliases to identify a particular action. Each alias points to action factory name, its configuration.
+If action is a behaviour, it points also to wrapped action (also by alias).
```
Comment (Voycawojka) [inline actions/README.md]:
```suggestion
Action provider initializes actions, combines actions with behaviours and caches stateful ones. It
uses aliases to identify a particular action. Each alias points to an action factory name and therefore its configuration.
If action is a behaviour, it also points to a wrapped action (by alias as well).
```
Don't know if `and therefore` is appropriate but I'd rephrase this sentence in some way to be more clear

File: actions/README.md [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+See the sample configuration below:
+```hocon
+actions {
+  ALIAS_A: ACTION_FACTORY_OPTIONS_A
+  ALIAS_B: ACTION_FACTORY_OPTIONS_B
+}
+```
+
+### Action factory options
+Action factory options contains:
```
Comment (Voycawojka) [inline actions/README.md]:
```suggestion
Action factory options contain:
```

File: actions/README.md (hunk 5 above)
Comment (tomaszmichalak) [inline actions/README.md]:
```suggestion
- retrieve the [action provider](#action-provider) to and invoke the action e.g.
```

File: actions/README.md (hunk 3 above)
Comment (tomaszmichalak) [inline actions/README.md]:
```suggestion
module as an example.
```

File: actions/README.md (hunk 4 above)
Comment (tomaszmichalak) [inline actions/README.md]:
```suggestion
More details [here](https://github.com/Knotx/knotx-fragments/blob/master/action/core/docs/asciidoc/dataobjects.adoc#actionfactoryoptions).
```
